src/main/java/com/lginterview/filter/ThrottlingFilter.java

```diff
@@ -0,0 +1,67 @@
+package com.lginterview.filter;
+
+import com.google.common.util.concurrent.RateLimiter;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import javax.servlet.*;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+@Component
+public class ThrottlingFilter implements Filter {
```

Comment:
Maybe i misunderstood the requirements but there was no specification that the limitation should be per client. Also as i understood it the limitation should affect only storage requests - not stats requests. That's the reason why i didn't use tomcat connection limits.

Reply:
[MS] Again, sorry about that, the requirement was not clear.

The idea was to implement request cache within the application order to reduce number of write requests to database. The requests are stored in-memory and flushed to the database every couple seconds.

In our case it would mean that we cache INIT, SAVE and FLUSH request and we call coursesRepository.saveAll(cachedRequest) every couple seconds to store it (of course we would first process requests before saving them to the DB).